File: backend/app/services/crawler_service.py
"""
Crawler service layer
Orchestrates website crawling using hybrid approach (Scrapy + Playwright fallback)
"""
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.models.business_profile import BusinessProfile, CrawlStatus
from app.models.website_content import WebsiteContent
from app.crawlers.playwright_crawler import run_playwright_crawler
from datetime import datetime
from typing import List, Dict, Optional
from loguru import logger
import asyncio
import json
from urllib.parse import urlparse


class CrawlerService:
    """Service for managing website crawling operations"""
    
    @staticmethod
    async def start_crawl(db: Session, business_profile_id: str) -> Dict:
        """
        Start website crawl for a business profile
        
        Args:
            db: Database session
            business_profile_id: ID of the business profile
            
        Returns:
            Dictionary with crawl start status
            
        Raises:
            ValueError: If business profile not found or has no website
        """
        # Get business profile
        business_profile = db.query(BusinessProfile).filter(
            BusinessProfile.id == business_profile_id
        ).first()
        
        if not business_profile:
            raise ValueError("Business profile not found")
        
        if not business_profile.website:
            raise ValueError("Business profile has no website URL")
        
        # Check if crawl already in progress
        if business_profile.crawl_status == CrawlStatus.IN_PROGRESS:
            return {
                "status": "already_in_progress",
                "message": "Crawl is already in progress for this business profile"
            }
        
        # Update crawl status to IN_PROGRESS
        business_profile.crawl_status = CrawlStatus.IN_PROGRESS
        business_profile.crawl_started_at = datetime.utcnow()
        business_profile.crawl_error = None
        db.commit()
        
        logger.info(f"Starting crawl for business profile {business_profile_id}: {business_profile.website}")
        
        try:
            # Use Playwright crawler (works for all sites including JS-heavy)
            pages_data = await run_playwright_crawler(
                start_url=business_profile.website,
                max_pages=30,
                max_depth=2
            )
            
            # Check if we got enough content
            if not pages_data or len(pages_data) == 0:
                raise Exception("No pages were successfully crawled")
            
            # Store extracted content in database
            stored_count = CrawlerService._store_crawled_content(
                db=db,
                business_profile_id=business_profile_id,
                pages_data=pages_data
            )
            
            # Update business profile with success status
            business_profile.crawl_status = CrawlStatus.COMPLETED
            business_profile.crawl_completed_at = datetime.utcnow()
            business_profile.total_pages_crawled = str(stored_count)
            db.commit()
            
            logger.info(f"Crawl completed for business profile {business_profile_id}: {stored_count} pages")
            
            return {
                "status": "completed",
                "message": f"Successfully crawled {stored_count} pages",
                "pages_crawled": stored_count
            }
        
        except Exception as e:
            logger.error(f"Crawl failed for business profile {business_profile_id}: {e}")
            
            # Update business profile with failed status
            business_profile.crawl_status = CrawlStatus.FAILED
            business_profile.crawl_error = str(e)
            business_profile.crawl_completed_at = datetime.utcnow()
            db.commit()
            
            return {
                "status": "failed",
                "message": f"Crawl failed: {str(e)}",
                "error": str(e)
            }
    
    @staticmethod
    def _store_crawled_content(
        db: Session,
        business_profile_id: str,
        pages_data: List[Dict]
    ) -> int:
        """
        Store crawled content in database
        
        Args:
            db: Database session
            business_profile_id: ID of the business profile
            pages_data: List of page data dictionaries
            
        Returns:
            Number of pages stored
        """
        stored_count = 0
        
        for page_data in pages_data:
            try:
                # Create WebsiteContent record
                content = WebsiteContent(
                    business_profile_id=business_profile_id,
                    url=page_data.get('url', ''),
                    page_type=page_data.get('page_type', 'other'),
                    raw_html=page_data.get('raw_html', ''),
                    cleaned_text=page_data.get('cleaned_text', ''),
                    title=page_data.get('title', ''),
                    meta_description=page_data.get('meta_description', ''),
                    h1_tags=json.dumps(page_data.get('h1_tags', [])),
                    h2_tags=json.dumps(page_data.get('h2_tags', [])),
                    extracted_name=page_data.get('extracted_name'),
                    extracted_address=page_data.get('extracted_address'),
                    extracted_phone=page_data.get('extracted_phone'),
                    word_count=page_data.get('word_count', 0),
                    crawled_at=datetime.utcnow()
                )
                
                db.add(content)
                stored_count += 1
            
            except Exception as e:
                logger.error(f"Error storing page content for {page_data.get('url')}: {e}")
                continue
        
        # Commit all content records
        db.commit()
        
        return stored_count
    
    @staticmethod
    def get_crawl_status(db: Session, business_profile_id: str) -> Dict:
        """
        Get crawl status for a business profile
        
        Args:
            db: Database session
            business_profile_id: ID of the business profile
            
        Returns:
            Dictionary with crawl status information
            
        Raises:
            ValueError: If business profile not found
        """
        # Get business profile
        business_profile = db.query(BusinessProfile).filter(
            BusinessProfile.id == business_profile_id
        ).first()
        
        if not business_profile:
            raise ValueError("Business profile not found")
        
        # Get page count
        page_count = db.query(func.count(WebsiteContent.id)).filter(
            WebsiteContent.business_profile_id == business_profile_id
        ).scalar()
        
        return {
            "status": business_profile.crawl_status.value,
            "started_at": business_profile.crawl_started_at.isoformat() if business_profile.crawl_started_at else None,
            "completed_at": business_profile.crawl_completed_at.isoformat() if business_profile.crawl_completed_at else None,
            "error": business_profile.crawl_error,
            "total_pages": int(business_profile.total_pages_crawled or "0"),
            "pages_in_db": page_count
        }
    
    @staticmethod
    def get_crawled_content(
        db: Session,
        business_profile_id: str,
        page_type: Optional[str] = None
    ) -> List[Dict]:
        """
        Get crawled content for a business profile
        
        Args:
            db: Database session
            business_profile_id: ID of the business profile
            page_type: Optional filter by page type (homepage, about, services, etc.)
            
        Returns:
            List of content dictionaries
            
        Raises:
            ValueError: If business profile not found
        """
        # Get business profile
        business_profile = db.query(BusinessProfile).filter(
            BusinessProfile.id == business_profile_id
        ).first()
        
        if not business_profile:
            raise ValueError("Business profile not found")
        
        # Build query
        query = db.query(WebsiteContent).filter(
            WebsiteContent.business_profile_id == business_profile_id
        )
        
        # Filter by page type if specified
        if page_type:
            query = query.filter(WebsiteContent.page_type == page_type)
        
        # Order by page type priority
        type_priority = {
            'homepage': 1,
            'about': 2,
            'services': 3,
            'products': 4,
            'contact': 5,
            'faq': 6,
            'other': 7
        }
        
        contents = query.all()
        
        # Sort by priority
        sorted_contents = sorted(
            contents,
            key=lambda x: type_priority.get(x.page_type, 999)
        )
        
        # Convert to dictionaries
        result = []
        for content in sorted_contents:
            result.append({
                "id": content.id,
                "url": content.url,
                "page_type": content.page_type,
                "title": content.title,
                "meta_description": content.meta_description,
                "h1_tags": json.loads(content.h1_tags) if content.h1_tags else [],
                "h2_tags": json.loads(content.h2_tags) if content.h2_tags else [],
                "cleaned_text": content.cleaned_text,
                "word_count": content.word_count,
                "extracted_name": content.extracted_name,
                "extracted_address": content.extracted_address,
                "extracted_phone": content.extracted_phone,
                "crawled_at": content.crawled_at.isoformat() if content.crawled_at else None
            })
        
        return result

File: backend/src/services/CrawlerService.ts
import { WebsiteContent, BusinessProfile } from '../models/index';
import { ValidationError, NotFoundError } from '../utils/errors';
import { Types } from 'mongoose';

export class CrawlerService {
  /**
   * Start website crawl for a business profile
   * In production, this would use Playwright/Puppeteer for actual crawling
   */
  static async startCrawl(profileId: string, websiteUrl: string): Promise<{ status: string; jobId: string }> {
    if (!profileId || !websiteUrl) {
      throw new ValidationError('profileId and websiteUrl are required');
    }

    // Validate URL format
    try {
      new URL(websiteUrl);
    } catch {
      throw new ValidationError('Invalid website URL format');
    }

    // Update business profile crawl status
    const profile = await BusinessProfile.findByIdAndUpdate(
      new Types.ObjectId(profileId),
      {
        crawlStatus: 'IN_PROGRESS',
        crawlStartedAt: new Date(),
        crawlError: null,
      },
      { new: true }
    );

    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }

    // Generate job ID
    const jobId = `job-${this.generateRandomString(16)}`;

    // Simulate crawl processing (in production, this would queue a job with Bull/Redis)
    // For now, we'll schedule a mock crawl to complete in a few seconds
    this.simulateCrawl(profileId, websiteUrl, jobId);

    return {
      status: 'PENDING',
      jobId,
    };
  }

  /**
   * Get crawl status for a profile
   */
  static async getCrawlStatus(
    profileId: string
  ): Promise<{
    status: string;
    totalPages: number;
    currentPage: number;
    progress: number;
    error?: string;
    startedAt?: Date;
    completedAt?: Date;
  }> {
    const profile = await BusinessProfile.findById(new Types.ObjectId(profileId));

    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }

    const pageCount = await WebsiteContent.countDocuments({
      businessProfileId: new Types.ObjectId(profileId),
    });

    const progress = profile.crawlStatus === 'IN_PROGRESS' ? 75 : profile.crawlStatus === 'COMPLETED' ? 100 : 0;

    return {
      status: profile.crawlStatus || 'NOT_STARTED',
      totalPages: pageCount,
      currentPage: Math.ceil(pageCount * 0.75), // Mock current progress
      progress,
      error: profile.crawlError || undefined,
      startedAt: profile.crawlStartedAt,
      completedAt: profile.crawlCompletedAt,
    };
  }

  /**
   * Get crawled content for a profile
   */
  static async getCrawledContent(profileId: string): Promise<any[]> {
    const pages = await WebsiteContent.find({
      businessProfileId: new Types.ObjectId(profileId),
    }).limit(100);

    return pages.map((page: any) => ({
      _id: page._id,
      url: page.url,
      title: page.title,
      description: page.metaDescription,
      content: page.cleanedText?.substring(0, 500), // Return first 500 chars
      headings: [page.h1Tags, page.h2Tags].filter(Boolean), // Use actual h1/h2 tags
      metadata: { wordCount: page.wordCount }, // Use actual metadata
      crawledAt: page.crawledAt,
    }));
  }

  /**
   * Extract metadata from HTML content
   */
  static async extractMetadata(
    html: string
  ): Promise<{
    title?: string;
    description?: string;
    keywords?: string;
    ogImage?: string;
  }> {
    // Mock metadata extraction
    // In production, use cheerio or similar HTML parser
    const titleMatch = html.match(/<title[^>]*>([^<]+)<\/title>/i);
    const descMatch = html.match(/<meta\s+name=["']description["'][^>]*content=["']([^"']+)["']/i);
    const keywordsMatch = html.match(/<meta\s+name=["']keywords["'][^>]*content=["']([^"']+)["']/i);
    const ogImageMatch = html.match(/<meta\s+property=["']og:image["'][^>]*content=["']([^"']+)["']/i);

    return {
      title: titleMatch ? titleMatch[1] : undefined,
      description: descMatch ? descMatch[1] : undefined,
      keywords: keywordsMatch ? keywordsMatch[1] : undefined,
      ogImage: ogImageMatch ? ogImageMatch[1] : undefined,
    };
  }

  /**
   * Simulate crawl completion (mock implementation)
   * In production, this would be handled by a separate job queue worker
   */
  private static simulateCrawl(profileId: string, websiteUrl: string, jobId: string): void {
    // Simulate crawl delay and then store mock pages
    setTimeout(async () => {
      try {
        // Create mock crawled pages
        const mockPages = [
          {
            url: websiteUrl,
            title: 'Home Page',
            metaDescription: 'Welcome to our business website',
            cleanedText: 'This is the main landing page with information about our services and products.',
            h1Tags: 'Welcome to Our Business',
            pageType: 'homepage',
          },
          {
            url: `${websiteUrl}/about`,
            title: 'About Us',
            metaDescription: 'Learn more about our company',
            cleanedText: 'We are a leading company in our industry with over 10 years of experience.',
            h1Tags: 'About Our Company',
            pageType: 'about',
          },
          {
            url: `${websiteUrl}/services`,
            title: 'Our Services',
            metaDescription: 'Explore our comprehensive service offerings',
            cleanedText: 'We offer a wide range of services tailored to meet your business needs.',
            h1Tags: 'Service Offerings',
            pageType: 'services',
          },
          {
            url: `${websiteUrl}/contact`,
            title: 'Contact Us',
            metaDescription: 'Get in touch with our team',
            cleanedText: 'Have questions? Contact our team for more information and support.',
            h1Tags: 'Contact Information',
            pageType: 'contact',
          },
        ];

        // Store crawled pages
        for (const mockPage of mockPages) {
          await WebsiteContent.create({
            businessProfileId: new Types.ObjectId(profileId),
            ...mockPage,
            crawledAt: new Date(),
          });
        }

        // Update profile crawl status
        await BusinessProfile.findByIdAndUpdate(new Types.ObjectId(profileId), {
          crawlStatus: 'COMPLETED',
          crawlCompletedAt: new Date(),
          totalPagesCrawled: mockPages.length,
          crawlError: null,
        });
      } catch (error) {
        // Update profile with error
        await BusinessProfile.findByIdAndUpdate(new Types.ObjectId(profileId), {
          crawlStatus: 'FAILED',
          crawlError: error instanceof Error ? error.message : 'Crawl failed',
        }).catch(() => {
          // Ignore update errors
        });
      }
    }, 3000); // Simulate 3-second crawl delay
  }

  /**
   * Helper: Generate random string
   */
  private static generateRandomString(length: number): string {
    const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    let result = '';
    for (let i = 0; i < length; i++) {
      result += chars.charAt(Math.floor(Math.random() * chars.length));
    }
    return result;
  }
}

export default CrawlerService;
